source/common/stats/symbol_table_impl.h

```diff
@@ -85,54 +211,172 @@ class SymbolTableImpl : public SymbolTable {
+class StatNameStorage {
 public:
-  StatNameImpl(SymbolVec symbol_vec, SymbolTableImpl& symbol_table)
-      : symbol_vec_(symbol_vec), symbol_table_(symbol_table) {}
-  ~StatNameImpl() override { symbol_table_.free(symbol_vec_); }
-  std::string toString() const override { return symbol_table_.decode(symbol_vec_); }
+  StatNameStorage(absl::string_view name, SymbolTable& table);
+  StatNameStorage(StatNameStorage&& src) : bytes_(std::move(src.bytes_)) {}
+
+  /**
+   * Before allowing a StatNameStorage to be destroyed, you must call free()
+   * on it, to drop the references to the symbols, allowing the SymbolTable
+   * to shrink.
+   */
+  ~StatNameStorage();
+
+  /**
+   * Decremences the reference counts in the SymbolTable.
+   *
+   * @param table the symbol table.
+   */
+  void free(SymbolTable& table);
+
+  /**
+   * @return StatName a reference to the owned storage.
+   */
+  inline StatName statName() const; // implementation below.
 
 private:
+  std::unique_ptr<SymbolStorage> bytes_;
+};
+
+/**
+ * Efficiently represents a stat name using a variable-length array of uint8_t.
+ * This class does not own the backing store for this array; the backing-store
+ * can be held in StatNameStorage, or it can be packed more tightly into another
+ * object.
+ *
+ * For large numbers of clusters, there are a huge number of StatNames so
+ * avoiding extra per-stat pointers has a significant memory impact.
+ */
+class StatName {
+public:
+  explicit StatName(const SymbolStorage symbol_array) : symbol_array_(symbol_array) {}
+  StatName() : symbol_array_(nullptr) {}
+
+  std::string toString(const SymbolTable& table) const { return table.decode(data(), numBytes()); }
+
+  /**
+   * Note that this hash function will return a different hash than that of
+   * the elaborated string.
+   *
+   * @return uint64_t a hash of the underlying symbol vector, since StatNames
+   *                  are uniquely defined by their symbol vectors.
+   */
+  uint64_t hash() const {
+    const char* cdata = reinterpret_cast<const char*>(data());
+    return HashUtil::xxHash64(absl::string_view(cdata, numBytes()));
+  }
+
+  // Compares on the underlying symbol vectors.
+  // NB: operator==(std::vector) checks size first, then compares equality for each element.
+  bool operator==(const StatName& rhs) const {
+    const size_t sz = numBytes();
+    return sz == rhs.numBytes() && memcmp(data(), rhs.data(), sz * sizeof(uint8_t)) == 0;
+  }
+  bool operator!=(const StatName& rhs) const { return !(*this == rhs); }
+
+protected:
+  friend SymbolTable;
   friend class StatNameTest;
-  SymbolVec symbolVec() { return symbol_vec_; }
-  SymbolVec symbol_vec_;
-  SymbolTableImpl& symbol_table_;
+  friend class StatNameJoiner;
+
+  /**
+   * @return size_t the number of bytes in the symbol array, excluding the two-byte
+   *                 overhead for the size itself.
+   */
+  size_t numBytes() const {
+    return symbol_array_[0] | (static_cast<size_t>(symbol_array_[1]) << 8);
```

Comment:
Do we care about endian-ness here? Can we just cast/store the 2 bytes and then retrieve them? Might be simpler to read?